src/state/opamStateConfig.ml

```diff
@@ -281,12 +281,44 @@ module Repos = struct
       (OpamPath.repos_config gt.root)
 end
 
+let downgrade_2_1_switch f =
+  let filename = OpamFile.filename f in
+  let str_f = OpamFilename.to_string filename in
+  let opamfile = OpamParser.FullPos.file str_f in
+  let opamfile' =
+    let open OpamParserTypes.FullPos in
+    { opamfile with
+      file_contents =
+        List.map (fun item ->
+            match item.pelem with
+            | Variable (({pelem = "opam-version"; _} as opam_v),
+                        ({pelem = String "2.1"; _} as v)) ->
+              { item with
+                pelem = Variable ({opam_v with pelem = "opam-version"},
+                                  {v with pelem = String "2.0"})}
+            | _ -> item) opamfile.file_contents}
+  in
+  if opamfile' = opamfile then None else
+    Some (opamfile'
+          |> OpamPrinter.FullPos.opamfile
+          |> OpamFile.Switch_config.read_from_string)
+
 let local_switch_exists root switch =
   (* we don't use safe loading function to avoid errors displaying *)
-  OpamPath.Switch.switch_config root switch |>
-  OpamFile.Switch_config.BestEffort.read_opt |> function
+  let f = OpamPath.Switch.switch_config root switch in
+  match OpamFile.Switch_config.BestEffort.read_opt f with
   | None -> false
   | Some conf -> conf.OpamFile.Switch_config.opam_root = Some root
+  | exception (OpamPp.Bad_version _ as e) ->
+    match OpamFile.Config.raw_root_version (OpamPath.config root) with
+    | None -> raise e
+    | Some _ ->
```

Comment:
Should it be constrained to root version `2.1~rc/2.1` ?

Reply:
I think so - doesn't this become a problem in future if it gets `opam-root-version: "2.2"`?

Reply:
With a `2.2` opam root, the local switch upgrade won't be triggered, yes.

Reply:
Ah, of course, the whole thing is in the exception for an incorrect version.

Reply:
Since this fix is only necessary for intermediate dev versions, I'd say we remove it from 2.2 or at least 2.3: it'll keep the upgrade code more maintainable and we can reasonably assume anyone who tried one of the affected versions switched to the release in the meantime ?

Reply:
it's unregistered local switch, they can stay around for a while without noticing